odoo/tools/safe_eval.py

```diff
@@ -1,225 +1,576 @@
+
+        return node
+
+    def explore_Subscript(self, node):
+        """
+        Explores every subscripts.
+        Wraps the subscript to the __SafeWrapper function, to check the type of the subscripted object.
+        We do this because we can't check the type of the subscripted object at compile time.
+        """
+        node = self.generic_visit(node)
+
+        return ast.Subscript(
+            value=ast.Call(
+                func=ast.Name("__SafeWrapper"),
+                args=[node.value, ast.Name(id="__type_checker")],
+                keywords=[]),
+            slice=node.slice,
+            ctx=node.ctx
+        )
+
+    def explore_Constant(self, node):
+        return super().visit_Constant(node)
+
+    def explore_Attribute(self, node):
+        """
+        Explores every attributes.
+        Wraps the node inside of the __SafeWrapper function, to check the type of the object dynamically.
+        """
+        node = self.generic_visit(node)
+
+        return ast.Attribute(
+            value=ast.Call(
+                func=ast.Name("__SafeWrapper", ctx=ast.Load()),
+                args=[node.value, ast.Name(id="__type_checker")],
+                keywords=[]),
+            attr=node.attr,
+            ctx=ast.Load())
+
+    def get_environment(self):
+        """
+        :return: the environment of the sandbox
+        """
+
+        __allowed_types = tuple(_ALLOWED_TYPES) + self.allow_type
+        __allowed_instances = tuple(_ALLOWED_INSTANCES) + self.allow_instance + self.allow_type
+
+        def type_checker(obj):
+            """Checks if the object is safe to be used in eval.
+            If it's not, it raises a :exc:`TypeError`.
+
+            :param obj: The object to check.
+
+            :return: The object if it is safe to be used in eval.
+            """
+            if callable(obj) and (obj is safe_eval or obj is eval or obj is exec): # pylint: disable=comparison-with-callable
+                raise TypeError("You cannot pass the safe_eval function to any objects")
+
+            if isinstance(obj, __allowed_instances) or (type(obj) is type and obj in __allowed_types): # pylint: disable=unidiomatic-typecheck
+                return obj
+            elif isinstance(obj, _wrapped_instances) or (type(obj) is type and obj in _wrapped_types) or (type(obj) is types.ModuleType and obj.__name__ in _ALLOWED_MODULES): # noqa: E721 (Reason: `type(...) is ...` is stricter than `isinstance(..., ...)` in this case) # pylint: disable=unidiomatic-typecheck
+                return SafeWrapper(obj, type_checker)
+            else:
+                raise TypeError(
+                    f"Object {obj} of type '{type(obj)}' is not allowed."
+                )
+
+        def call_checker(func, *args, **kwargs):
+            if func in [safe_eval, exec, eval]: # pylint: disable=comparison-with-callable
+                raise TypeError("You cannot pass the safe_eval function to any objects")
```

Comment:
I don't see why checking the `__self__` will help us in this case.

Reply:
because `__self__` will be `builtins`, so we can block that
(and replacing len with a lambda for example will remove that `__self__`, so it will be accepted)

Reply:
What should I do in this case ?

```python
lambda x: exec(x)
```

Reply:
it will call explore_Call with exec no ? (when you call the lambda first, then when you call exec because it's inside the safe eval)
so it'll be check no ?